File: zertz-core/union_find.zig
const std = @import("std");
const sort = std.sort;

const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const AutoHashMap = std.AutoHashMap;

pub fn UnionFind(comptime T: type) type {
    return struct {
        allocator: Allocator,
        inner: AutoHashMap(T, usize),
        reverse: AutoHashMap(usize, T),
        parent_data: ArrayList(usize),
        rank_data: ArrayList(usize),

        const Self = @This();

        pub fn init(allocator: Allocator, elems: []const T) !Self {
            var output: Self = undefined;

            output.allocator = allocator;
            output.inner = AutoHashMap(T, usize).init(allocator);
            errdefer output.inner.deinit();

            output.reverse = AutoHashMap(usize, T).init(allocator);
            errdefer output.reverse.deinit();

            output.parent_data = try ArrayList(usize).initCapacity(allocator, elems.len);
            errdefer output.parent_data.deinit();

            output.rank_data = try ArrayList(usize).initCapacity(allocator, elems.len);
            errdefer output.rank_data.deinit();

            // Rreserve both HashMap capacity enough to use
            try output.inner.ensureTotalCapacity(@intCast(u32, elems.len));
            try output.reverse.ensureTotalCapacity(@intCast(u32, elems.len));

            for (elems) |elem, idx| {
                try output.inner.put(elem, idx);
                try output.reverse.put(idx, elem);
                try output.parent_data.append(idx);
                try output.rank_data.append(0);
            }

            return output;
        }

        pub fn deinit(self: *Self) void {
            self.inner.deinit();
            self.reverse.deinit();
            self.parent_data.deinit();
            self.rank_data.deinit();
        }

        pub fn unionBoth(self: *Self, x: T, y: T) void {
            const x_location = self.find(x);
            const y_location = self.find(y);

            if (x_location == y_location) {
                return;
            }

            if (self.rank_data.items[x_location] < self.rank_data.items[y_location]) {
                self.parent_data.items[x_location] = y_location;
            } else {
                self.parent_data.items[y_location] = x_location;

                if (self.rank_data.items[x_location] == self.rank_data.items[y_location]) {
                    self.rank_data.items[x_location] += 1;
                }
            }
        }

        pub fn find(self: *Self, x: T) usize {
            const x_location = self.inner.get(x).?;
            return self.findParent(x_location);
        }

        pub fn clear(self: *Self) void {
            var key_iter = self.inner.keyIterator();

            var idx: usize = 0;
            while (key_iter.next()) |elem| {
                self.inner.getPtr(elem.*).?.* = idx;
                self.reverse.getPtr(idx).?.* = elem.*;
                self.parent_data.items[idx] = idx;
                idx += 1;
            }

            for (self.rank_data.items) |*data| {
                data.* = 0;
            }
        }

        pub fn getComponentRepresentors(self: *Self) !ArrayList(T) {
            var parent_data = try self.parent_data.clone();
            defer parent_data.deinit();

            sort.sort(usize, parent_data.items, {}, comptime sort.asc(usize));

            var i: usize = 0;
            var j: usize = 0;
            while (j < parent_data.items.len) : (j += 1) {
                if (parent_data.items[i] == parent_data.items[j]) {
                    continue;
                }
                i += 1;
                parent_data.items[i] = parent_data.items[j];
            }
            parent_data.shrinkRetainingCapacity(i + 1);

            var output = try ArrayList(T).initCapacity(self.allocator, self.inner.count());
            errdefer output.deinit();

            for (parent_data.items) |pd| {
                try output.append(self.reverse.get(pd).?);
            }

            return output;
        }

        fn findParent(self: *Self, location: usize) usize {
            return if (self.parent_data.items[location] == location)
                location
            else blk: {
                const root_parent = self.findParent(self.parent_data.items[location]);
                self.parent_data.items[location] = root_parent;
                break :blk root_parent;
            };
        }
    };
}

// Testing UnionFind
const test_allocator = std.testing.allocator;
const expectEqual = std.testing.expectEqual;

test "test unionfind data structure" {
    const elems = [_]i32{ 1, 2, 3, 4, 5, 6, 7, 8 };
    var union_find = try UnionFind(i32).init(test_allocator, &elems);
    defer union_find.deinit();

    union_find.unionBoth(1, 2);
    union_find.unionBoth(4, 5);
    union_find.unionBoth(6, 1);
    union_find.unionBoth(3, 7);
    union_find.unionBoth(7, 8);
    union_find.unionBoth(2, 5);

    try expectEqual(union_find.find(1), union_find.find(6));
    try expectEqual(union_find.find(2), union_find.find(6));
    try expectEqual(union_find.find(3), union_find.find(3));
    try expectEqual(union_find.find(4), union_find.find(6));
    try expectEqual(union_find.find(5), union_find.find(6));
    try expectEqual(union_find.find(6), union_find.find(6));
    try expectEqual(union_find.find(7), union_find.find(3));
    try expectEqual(union_find.find(8), union_find.find(3));

    union_find.clear();

    union_find.unionBoth(2, 3);
    union_find.unionBoth(1, 5);
    union_find.unionBoth(6, 7);
    union_find.unionBoth(7, 5);

    try expectEqual(union_find.find(1), union_find.find(1));
    try expectEqual(union_find.find(2), union_find.find(2));
    try expectEqual(union_find.find(3), union_find.find(2));
    try expectEqual(union_find.find(4), union_find.find(4));
    try expectEqual(union_find.find(5), union_find.find(1));
    try expectEqual(union_find.find(6), union_find.find(1));
    try expectEqual(union_find.find(7), union_find.find(1));
    try expectEqual(union_find.find(8), union_find.find(8));
}

File: zertz-core/src/union_find.rs
use std::collections::HashMap;
use std::hash::Hash;

#[derive(Debug, Clone, Default)]
pub struct UnionFind<T> {
    inner: HashMap<T, usize>,
    reverse: HashMap<usize, T>,
    parent_data: Vec<usize>,
    rank_data: Vec<usize>,
}

impl<T> UnionFind<T>
where
    T: Eq + Hash,
{
    pub fn union(&mut self, x: &T, y: &T) {
        let x_location = self.find(x);
        let y_location = self.find(y);

        if x_location == y_location {
            return;
        }

        if self.rank_data[x_location] < self.rank_data[y_location] {
            self.parent_data[x_location] = y_location;
        } else {
            self.parent_data[y_location] = x_location;

            if self.rank_data[x_location] == self.rank_data[y_location] {
                self.rank_data[x_location] += 1;
            }
        }
    }

    pub fn find(&mut self, x: &T) -> usize {
        let x_location = self
            .inner
            .get(x)
            .expect("[Zertz Internal Error]: UnionFind::Find");
        self.find_parent(*x_location)
    }

    fn find_parent(&mut self, location: usize) -> usize {
        if self.parent_data[location] == location {
            location
        } else {
            let root_parent = self.find_parent(self.parent_data[location]);
            self.parent_data[location] = root_parent;
            root_parent
        }
    }
}

impl<T> UnionFind<T>
where
    T: Clone + Eq + Hash,
{
    pub fn clear(&mut self) {
        let keys: Vec<T> = self.inner.keys().cloned().collect();

        for (idx, elem) in keys.into_iter().enumerate() {
            *self.inner.get_mut(&elem).unwrap() = idx;
            *self.reverse.get_mut(&idx).unwrap() = elem;
            self.parent_data[idx] = idx;
        }
        for data in self.rank_data.iter_mut() {
            *data = 0;
        }
    }

    pub fn components(&self) -> Vec<T> {
        let mut output = Vec::with_capacity(self.inner.len());
        let mut parent_data = self.parent_data.clone();
        parent_data.sort();
        parent_data.dedup();

        for pd in parent_data {
            output.push(self.reverse[&pd].clone());
        }

        output
    }
}

impl<T> From<Vec<T>> for UnionFind<T>
where
    T: Clone + Eq + Hash,
{
    fn from(elems: Vec<T>) -> Self {
        let mut inner = HashMap::with_capacity(elems.len());
        let mut reverse = HashMap::with_capacity(elems.len());
        let mut parent_data = Vec::with_capacity(elems.len());
        let rank_data = vec![0; elems.len()];

        for (idx, elem) in elems.into_iter().enumerate() {
            inner.insert(elem.clone(), idx);
            reverse.insert(idx, elem);
            parent_data.push(idx);
        }

        Self {
            inner,
            reverse,
            parent_data,
            rank_data,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn one_item_union_find() {
        let elems = vec![1];
        let mut union_find = UnionFind::from(elems);

        assert_eq!(union_find.find(&1), 0);
    }

    #[test]
    fn complex_union_find() {
        let elems = vec![1, 2, 3, 4, 5, 6, 7, 8];
        let mut union_find = UnionFind::from(elems);

        union_find.union(&1, &2);
        union_find.union(&4, &5);
        union_find.union(&6, &1);
        union_find.union(&3, &7);
        union_find.union(&7, &8);
        union_find.union(&2, &5);

        assert_eq!(union_find.find(&1), union_find.find(&6));
        assert_eq!(union_find.find(&2), union_find.find(&6));
        assert_eq!(union_find.find(&3), union_find.find(&3));
        assert_eq!(union_find.find(&4), union_find.find(&6));
        assert_eq!(union_find.find(&5), union_find.find(&6));
        assert_eq!(union_find.find(&6), union_find.find(&6));
        assert_eq!(union_find.find(&7), union_find.find(&3));
        assert_eq!(union_find.find(&8), union_find.find(&3));

        union_find.clear();

        union_find.union(&2, &3);
        union_find.union(&1, &5);
        union_find.union(&6, &7);
        union_find.union(&7, &5);

        assert_eq!(union_find.find(&1), union_find.find(&1));
        assert_eq!(union_find.find(&2), union_find.find(&2));
        assert_eq!(union_find.find(&3), union_find.find(&2));
        assert_eq!(union_find.find(&4), union_find.find(&4));
        assert_eq!(union_find.find(&5), union_find.find(&1));
        assert_eq!(union_find.find(&6), union_find.find(&1));
        assert_eq!(union_find.find(&7), union_find.find(&1));
        assert_eq!(union_find.find(&8), union_find.find(&8));
    }
}
